### PR TITLE
High Level Functions

### PR DESCRIPTION
**Pull request by samuel-tate**
## Issues:
The main issue this relates to is:
AlexsLemonade/refinebio#2545

This PR also takes care of:
#9 

## Also:
This PR also adds `PaginatedList` which can be found in `util.py`

Now, all `foo.search()` methods that deal with paginated api responses will return a `PaginatedList` of results.

Previously these methods would return generators which was fine, but sometimes when testing I found that I wanted to get the first object or index the results. This was kinda annoying with generators. `PaginatedList` can be indexed and also can be iterated like a generator.

### CODE REVIEW

File: pyrefinebio/__init__.py [hunk 1]
```diff
@@ -22,3 +22,9 @@
     NotFound,
     DownloadError
 )
+
+from pyrefinebio.hlf import (
+    help,
+    download_compendium,
+    download_quandfile_compendium
+)
```
Comment (samuel-tate) [inline pyrefinebio/__init__.py:30]: newline!

File: pyrefinebio/hlf.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        
+        gettting info about class methods:
+
+        >>> pyrefinebio.help("Sample.get")
+        or
+        >>> pyrefinebio.help("Sample get")
+    """
+
+    if not entity:
+        print()
```
Comment (samuel-tate) [inline pyrefinebio/hlf.py:29]: wasn't sure what to print as a high level overview

File: pyrefinebio/__init__.py [hunk 3]
```diff
@@ -22,3 +22,9 @@
     NotFound,
     DownloadError
 )
+
+from pyrefinebio.hlf import (
```
Comment (kurtwheeler) [inline pyrefinebio/__init__.py:26]: I don't love having an abbreviation as a module name. I think spelling out `high_level_functions` is more verbose but 
far more clear for new contributors.

File: pyrefinebio/dataset.py [hunk 4]
```diff
@@ -61,10 +61,24 @@ def __init__(
...
         return Dataset(**response)
 
 
+    def add_samples(self, experiment, samples=["ALL"]):
+        if self.is_processing or self.is_processed:
+            print("Cannot add samples to a dataset that has been processed!")
+            return
+
+        if self.data:
+            self.data = {**self.data, experiment: samples}
```
Comment (kurtwheeler) [inline pyrefinebio/dataset.py:74]: Any reason to do this instead of `self.data[experiment] = samples`?

Also are `experiment` and `samples` objects or strings of accession codes here? It might actually make things easier to use if we support both seamlessly. (The dict we POST will need to be all strings though.)

File: pyrefinebio/token.py [hunk 5]
```diff
@@ -48,9 +48,7 @@ def agree_to_terms_and_conditions(cls, api_token):
             api_token (str): the uuid string identifying the token
                              you want to activate.
         """
-
-        response = put_by_endpoint("token/" + api_token, payload={"is_activated": True})
-        return response
+        put_by_endpoint("token/" + api_token, payload={"is_activated": True})
```
Comment (kurtwheeler) [inline pyrefinebio/token.py:51]:
```suggestion
        return put_by_endpoint("token/" + api_token, payload={"is_activated": True})
```
I think returning the response here could be helpful in understanding what went wrong.

File: pyrefinebio/util.py [hunk 6]
```diff
@@ -2,21 +2,82 @@
...
+        return self.total_items
+
+
+    def __next__(self):
+        if self.cur < self.total_items:
+            n = self[self.cur]
+            self.cur += 1
+            return n
+        else:
+            raise StopIteration()
```
Comment (kurtwheeler) [inline pyrefinebio/util.py:83]: As far as I can tell, this class looks to be well implemented! I think it might be worth making some unit tests for it though to make sure there's no edge cases or off by one errors lurking.

File: pyrefinebio/hlf.py [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+        getting info about classes:
+
+        >>> pyrefinebio.help("Sample")
+        
+        gettting info about class methods:
+
+        >>> pyrefinebio.help("Sample.get")
+        or
+        >>> pyrefinebio.help("Sample get")
```
Comment (kurtwheeler) [inline pyrefinebio/hlf.py:25]: This works for functions as well as methods right? Could we include an example of that:  `pyrefinebio.help("download_dataset")`

**Review by kurtwheeler — CHANGES_REQUESTED** [review]
This looks great!! I left some comments, we discussed the high level overview and what to do about constructing datasets on a call, and I think you're aware that the new functions need docstrings.

File: pyrefinebio/dataset.py [hunk 8]
```diff
@@ -82,15 +84,18 @@ def add_samples(self, experiment, samples=["ALL"]):
             print("Cannot add samples to a dataset that has been processed!")
             return
 
+        # get accession codes if Experiment or Sample objects are passed in
+        experiment = experiment.accession_code if isinstance(experiment, prb_experiment.Experiment) else experiment
+        samples = [sample.accession_code if isinstance(sample, prb_sample.Sample) else sample for sample in samples]
```
Comment (kurtwheeler) [inline pyrefinebio/dataset.py:89]: Sweet, can we mention this in the docstring?

**Review by kurtwheeler — APPROVED** [review]
This looks great! I just had one small comment I think!
